{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "87f427a4-97e2-4901-8371-2077aa2a4e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from matplotlib import pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import glob\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b4243a6-7759-47d1-a298-9d85a5c54d05",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_table(\"/mnt/nfs/wangd/project/paper_db/peps_all_info_to_str.tsv\")\n",
    "res[\"pep_id\"] = res.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abc1609f",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_split = res['all_info'].str.split(';', expand=True).stack().reset_index(level=1, drop=True).to_frame('all_info')\n",
    "res = res.drop('all_info', axis=1).join(res_split).reset_index(drop=True)\n",
    "\n",
    "res_split2 = res['all_info'].str.split(',', expand=True)\n",
    "res_split2.columns = ['sample', 'protein', 'gene', 'transcript', 'gene_biotype', 'gene_symbol']\n",
    "res = pd.concat([res, res_split2], axis=1)\n",
    "res = res.drop('all_info', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "75cc74b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "res[\"gtf_index\"] = res[\"gene\"].str.split(\".\").apply(lambda x: x[0]) + \",\" + res[\"gene_symbol\"] + \",\" + res[\"gene_biotype\"]\n",
    "res[\"xref_index\"] = res[\"gene\"].str.split(\".\").apply(lambda x: x[0]) + \",\" + res[\"transcript\"].str.split(\".\").apply(lambda x: x[0]) + \",\" + res[\"protein\"].str.split(\".\").apply(lambda x: x[0])\n",
    "res.drop_duplicates()\n",
    "res.fillna(\"\",inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799e6984",
   "metadata": {},
   "source": [
    "### add gtf info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "82fa3ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_gtf(file):\n",
    "    gtf = pd.read_table(file, sep='\\t', comment='#', header=None, names=['chrom', 'source','feature','start','end','score ','strand','frame','attribute'], low_memory=False)\n",
    "    \n",
    "    gtf['gene_id'] = gtf['attribute'].str.extract('gene_id \"(.*?)\";')\n",
    "    gtf['gene_version'] = gtf['attribute'].str.extract('gene_version \"(.*?)\";')\n",
    "    gtf['gene_name'] = gtf['attribute'].str.extract('gene_name \"(.*?)\";')\n",
    "    gtf['gene_source'] = gtf['attribute'].str.extract('gene_source \"(.*?)\";')\n",
    "    gtf['gene_biotype'] = gtf['attribute'].str.extract('gene_biotype \"(.*?)\";')\n",
    "    gtf['projection_parent_gene'] = gtf['attribute'].str.extract('projection_parent_gene \"(.*?)\";')\n",
    "\n",
    "    gtf.fillna(\"\",inplace=True)\n",
    "    gtf.index = gtf[\"gene_id\"] + \",\" + gtf[\"gene_name\"] + \",\" + gtf[\"gene_biotype\"]\n",
    "\n",
    "    gtf.drop(['attribute', 'gene_id', 'gene_name', 'gene_biotype'], axis=1, inplace=True)\n",
    "    \n",
    "    gtf_dict = {}\n",
    "    for index, row in gtf.iterrows():\n",
    "        if index in gtf_dict:\n",
    "            gtf_dict[index].append( ','.join(map(str, row)))\n",
    "        else:\n",
    "            gtf_dict[index] = [ ','.join(map(str, row))]\n",
    "    \n",
    "    return gtf_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02b370e8",
   "metadata": {},
   "source": [
    "### add xref info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df3f041d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_xref(file):\n",
    "    xref = pd.read_table(file)\n",
    "\n",
    "    xref.fillna(\"\",inplace=True)\n",
    "    xref.index = xref[\"gene_stable_id\"] + \",\" + xref[\"transcript_stable_id\"] + \",\" + xref[\"protein_stable_id\"]\n",
    "    xref.drop(['gene_stable_id', 'transcript_stable_id', 'protein_stable_id'], axis=1, inplace=True)\n",
    "    \n",
    "    xref_dict = {}\n",
    "    for index, row in xref.iterrows():\n",
    "        if index in xref_dict:\n",
    "            xref_dict[index].append( ','.join(map(str, row)))\n",
    "        else:\n",
    "            xref_dict[index] = [ ','.join(map(str, row))]\n",
    "    return xref_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90e4a479",
   "metadata": {},
   "source": [
    "### add gtf And xref info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e963e20",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_info(info_dict,index):\n",
    "    info = info_dict.get(index)\n",
    "    if info:\n",
    "        return ';'.join(info) \n",
    "    else:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b937ca98",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 97/97 [5:12:46<00:00, 193.47s/it]  "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time consumption : 5:12:46.576471\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "start_time = datetime.datetime.now()\n",
    "\n",
    "all_dfs = []\n",
    "grouped_dfs = res.groupby(\"sample\")\n",
    "for sample, group_df in tqdm(grouped_dfs):\n",
    "    gtf_pattern = \"/mnt/nfs/wangd/project/paper_db/\" + \"gtf/*\" + sample + \"*\"\n",
    "    xref_pattern = \"/mnt/nfs/wangd/project/paper_db/\" + \"xref/*\" + sample + \"*\"\n",
    "    matching_gtf = glob.glob(gtf_pattern)[0]\n",
    "    matching_xref = glob.glob(xref_pattern)[0]\n",
    "    gtf_dict = add_gtf(matching_gtf)\n",
    "    xref_dict = add_xref(matching_xref)\n",
    "\n",
    "    group_df[\"gtf_info\"] = group_df[\"gtf_index\"].apply(lambda x: get_info(gtf_dict,x))\n",
    "    group_df[\"xref_info\"] = group_df[\"xref_index\"].apply(lambda x: get_info(xref_dict,x))\n",
    "\n",
    "    all_dfs.append(group_df)\n",
    "\n",
    "end_time = datetime.datetime.now()\n",
    "time_taken = end_time - start_time\n",
    "print(\"Time consumption :\", time_taken)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "40525975",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_res = pd.concat(all_dfs)\n",
    "all_res.to_csv(\"/mnt/nfs/wangd/project/paper_db/add_gtf_xref_info.tsv\", header=1, sep=\"\\t\",index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "684e407f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>protein_accessions</th>\n",
       "      <th>charge</th>\n",
       "      <th>scan_number</th>\n",
       "      <th>peptidoform</th>\n",
       "      <th>exp_mass_to_charge</th>\n",
       "      <th>calc_mass_to_charge</th>\n",
       "      <th>seq</th>\n",
       "      <th>tr</th>\n",
       "      <th>modifications</th>\n",
       "      <th>...</th>\n",
       "      <th>sample</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gtf_index</th>\n",
       "      <th>xref_index</th>\n",
       "      <th>gtf_info</th>\n",
       "      <th>xref_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>GCA_109298</td>\n",
       "      <td>2</td>\n",
       "      <td>28971</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>1172.107178</td>\n",
       "      <td>1171.600967</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>4215.9185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_009914755.4</td>\n",
       "      <td>ENSP05220080001.1</td>\n",
       "      <td>ENSG05220046501.1</td>\n",
       "      <td>ENST05220179403.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SRP14</td>\n",
       "      <td>ENSG05220046501,SRP14,protein_coding</td>\n",
       "      <td>ENSG05220046501,ENST05220179403,ENSP05220080001</td>\n",
       "      <td>15,ensembl,gene,37841027,37844451,.,-,.,1,ense...</td>\n",
       "      <td>GO:0005786,GO:0005786,signal recognition parti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>GCA_109298</td>\n",
       "      <td>2</td>\n",
       "      <td>28971</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>1172.107178</td>\n",
       "      <td>1171.600967</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>4215.9185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_009914755.4</td>\n",
       "      <td>ENSP05220080004.1</td>\n",
       "      <td>ENSG05220046501.1</td>\n",
       "      <td>ENST05220179408.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SRP14</td>\n",
       "      <td>ENSG05220046501,SRP14,protein_coding</td>\n",
       "      <td>ENSG05220046501,ENST05220179408,ENSP05220080004</td>\n",
       "      <td>15,ensembl,gene,37841027,37844451,.,-,.,1,ense...</td>\n",
       "      <td>GO:0005786,GO:0005786,signal recognition parti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>GCA_109298</td>\n",
       "      <td>2</td>\n",
       "      <td>28971</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>1172.107178</td>\n",
       "      <td>1171.600967</td>\n",
       "      <td>AAAAAAAAAPAAAATAATTAATTAATAAQ</td>\n",
       "      <td>4215.9185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_009914755.4</td>\n",
       "      <td>ENSP05220080007.1</td>\n",
       "      <td>ENSG05220046501.1</td>\n",
       "      <td>ENST05220179411.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SRP14</td>\n",
       "      <td>ENSG05220046501,SRP14,protein_coding</td>\n",
       "      <td>ENSG05220046501,ENST05220179411,ENSP05220080007</td>\n",
       "      <td>15,ensembl,gene,37841027,37844451,.,-,.,1,ense...</td>\n",
       "      <td>GO:0005786,GO:0005786,signal recognition parti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAFYKNILGAQVSEVVPLPEHGVSVVFVNLGNTK</td>\n",
       "      <td>GCA_133924</td>\n",
       "      <td>4</td>\n",
       "      <td>62173</td>\n",
       "      <td>AAAFYKNILGAQ[Deamidated]VSEVVPLPEHGVSVVFVNLGNTK</td>\n",
       "      <td>918.255432</td>\n",
       "      <td>917.997627</td>\n",
       "      <td>AAAFYKNILGAQVSEVVPLPEHGVSVVFVNLGNTK</td>\n",
       "      <td>5808.4260</td>\n",
       "      <td>12|Deamidated</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_009914755.4</td>\n",
       "      <td>ENSP05220077422.1</td>\n",
       "      <td>ENSG05220045157.1</td>\n",
       "      <td>ENST05220173529.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MCEE</td>\n",
       "      <td>ENSG05220045157,MCEE,protein_coding</td>\n",
       "      <td>ENSG05220045157,ENST05220173529,ENSP05220077422</td>\n",
       "      <td>2,ensembl,gene,71120697,71141279,.,-,.,1,ensem...</td>\n",
       "      <td>UPI0000369399,UPI0000369399,,UniParc,CHECKSUM,,,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAASHLFPFEK</td>\n",
       "      <td>GCA_27750,GCA_107325,GCA_279617,GCA_313537</td>\n",
       "      <td>3</td>\n",
       "      <td>24015</td>\n",
       "      <td>AAASHLFPFEK</td>\n",
       "      <td>406.549468</td>\n",
       "      <td>406.548631</td>\n",
       "      <td>AAASHLFPFEK</td>\n",
       "      <td>2901.2000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_009914755.4</td>\n",
       "      <td>ENSP05220065711.1</td>\n",
       "      <td>ENSG05220038055.1</td>\n",
       "      <td>ENST05220146556.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>XPC</td>\n",
       "      <td>ENSG05220038055,XPC,protein_coding</td>\n",
       "      <td>ENSG05220038055,ENST05220146556,ENSP05220065711</td>\n",
       "      <td>3,ensembl,gene,14147095,14180485,.,-,.,1,ensem...</td>\n",
       "      <td>A0A024R2M8,A0A024R2M8,,UniProtKB generic acces...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912085</th>\n",
       "      <td>YVQSAQSQIHNTCWAMMGLMAVR</td>\n",
       "      <td>GCA_102815</td>\n",
       "      <td>3</td>\n",
       "      <td>22873</td>\n",
       "      <td>YVQSAQSQ[Deamidated]IHNTC[Carbamidomethyl]WAMM...</td>\n",
       "      <td>900.753479</td>\n",
       "      <td>900.411420</td>\n",
       "      <td>YVQSAQSQIHNTCWAMMGLMAVR</td>\n",
       "      <td>3952.7363</td>\n",
       "      <td>8|Deamidated|13|Carbamidomethyl|17|Oxidation</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_021951015.1</td>\n",
       "      <td>ENSP05525043594.1</td>\n",
       "      <td>ENSG05525025770.1</td>\n",
       "      <td>ENST05525097559.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LSS</td>\n",
       "      <td>ENSG05525025770,LSS,protein_coding</td>\n",
       "      <td>ENSG05525025770,ENST05525097559,ENSP05525043594</td>\n",
       "      <td>21,ensembl,gene,38336946,38377856,.,-,.,1,ense...</td>\n",
       "      <td>GO:0005811,GO:0005811,lipid droplet,GO,DEPENDE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912086</th>\n",
       "      <td>YVSLENQHCEHCNSCTSK</td>\n",
       "      <td>GCA_156290</td>\n",
       "      <td>3</td>\n",
       "      <td>5609</td>\n",
       "      <td>YVSLENQHC[Carbamidomethyl]EHC[Carbamidomethyl]...</td>\n",
       "      <td>751.640869</td>\n",
       "      <td>751.642109</td>\n",
       "      <td>YVSLENQHCEHCNSCTSK</td>\n",
       "      <td>823.0072</td>\n",
       "      <td>9|Carbamidomethyl|12|Carbamidomethyl|15|Carbam...</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_021951015.1</td>\n",
       "      <td>ENSP05525070939.1</td>\n",
       "      <td>ENSG05525041216.1</td>\n",
       "      <td>ENST05525158826.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZCCHC4</td>\n",
       "      <td>ENSG05525041216,ZCCHC4,protein_coding</td>\n",
       "      <td>ENSG05525041216,ENST05525158826,ENSP05525070939</td>\n",
       "      <td>4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...</td>\n",
       "      <td>GO:0003676,GO:0003676,nucleic acid binding,GO,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912087</th>\n",
       "      <td>YVSLENQHCEHCNSCTSK</td>\n",
       "      <td>GCA_156290</td>\n",
       "      <td>3</td>\n",
       "      <td>5609</td>\n",
       "      <td>YVSLENQHC[Carbamidomethyl]EHC[Carbamidomethyl]...</td>\n",
       "      <td>751.640869</td>\n",
       "      <td>751.642109</td>\n",
       "      <td>YVSLENQHCEHCNSCTSK</td>\n",
       "      <td>823.0072</td>\n",
       "      <td>9|Carbamidomethyl|12|Carbamidomethyl|15|Carbam...</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_021951015.1</td>\n",
       "      <td>ENSP05525070944.1</td>\n",
       "      <td>ENSG05525041216.1</td>\n",
       "      <td>ENST05525158838.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZCCHC4</td>\n",
       "      <td>ENSG05525041216,ZCCHC4,protein_coding</td>\n",
       "      <td>ENSG05525041216,ENST05525158838,ENSP05525070944</td>\n",
       "      <td>4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...</td>\n",
       "      <td>GO:0003676,GO:0003676,nucleic acid binding,GO,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912088</th>\n",
       "      <td>YVSLENQHCEHCNSCTSKDGR</td>\n",
       "      <td>GCA_156290</td>\n",
       "      <td>5</td>\n",
       "      <td>9951</td>\n",
       "      <td>YVSLENQ[Deamidated]HC[Carbamidomethyl]EHC[Carb...</td>\n",
       "      <td>517.421082</td>\n",
       "      <td>517.214883</td>\n",
       "      <td>YVSLENQHCEHCNSCTSKDGR</td>\n",
       "      <td>1444.7499</td>\n",
       "      <td>7|Deamidated|9|Carbamidomethyl|12|Carbamidomet...</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_021951015.1</td>\n",
       "      <td>ENSP05525070939.1</td>\n",
       "      <td>ENSG05525041216.1</td>\n",
       "      <td>ENST05525158826.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZCCHC4</td>\n",
       "      <td>ENSG05525041216,ZCCHC4,protein_coding</td>\n",
       "      <td>ENSG05525041216,ENST05525158826,ENSP05525070939</td>\n",
       "      <td>4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...</td>\n",
       "      <td>GO:0003676,GO:0003676,nucleic acid binding,GO,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912089</th>\n",
       "      <td>YVSLENQHCEHCNSCTSKDGR</td>\n",
       "      <td>GCA_156290</td>\n",
       "      <td>5</td>\n",
       "      <td>9951</td>\n",
       "      <td>YVSLENQ[Deamidated]HC[Carbamidomethyl]EHC[Carb...</td>\n",
       "      <td>517.421082</td>\n",
       "      <td>517.214883</td>\n",
       "      <td>YVSLENQHCEHCNSCTSKDGR</td>\n",
       "      <td>1444.7499</td>\n",
       "      <td>7|Deamidated|9|Carbamidomethyl|12|Carbamidomet...</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_021951015.1</td>\n",
       "      <td>ENSP05525070944.1</td>\n",
       "      <td>ENSG05525041216.1</td>\n",
       "      <td>ENST05525158838.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZCCHC4</td>\n",
       "      <td>ENSG05525041216,ZCCHC4,protein_coding</td>\n",
       "      <td>ENSG05525041216,ENST05525158838,ENSP05525070944</td>\n",
       "      <td>4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...</td>\n",
       "      <td>GO:0003676,GO:0003676,nucleic acid binding,GO,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>912090 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 sequence_x  \\\n",
       "0             AAAAAAAAAPAAAATAATTAATTAATAAQ   \n",
       "1             AAAAAAAAAPAAAATAATTAATTAATAAQ   \n",
       "2             AAAAAAAAAPAAAATAATTAATTAATAAQ   \n",
       "3       AAAFYKNILGAQVSEVVPLPEHGVSVVFVNLGNTK   \n",
       "4                               AAASHLFPFEK   \n",
       "...                                     ...   \n",
       "912085              YVQSAQSQIHNTCWAMMGLMAVR   \n",
       "912086                   YVSLENQHCEHCNSCTSK   \n",
       "912087                   YVSLENQHCEHCNSCTSK   \n",
       "912088                YVSLENQHCEHCNSCTSKDGR   \n",
       "912089                YVSLENQHCEHCNSCTSKDGR   \n",
       "\n",
       "                                protein_accessions  charge  scan_number  \\\n",
       "0                                       GCA_109298       2        28971   \n",
       "1                                       GCA_109298       2        28971   \n",
       "2                                       GCA_109298       2        28971   \n",
       "3                                       GCA_133924       4        62173   \n",
       "4       GCA_27750,GCA_107325,GCA_279617,GCA_313537       3        24015   \n",
       "...                                            ...     ...          ...   \n",
       "912085                                  GCA_102815       3        22873   \n",
       "912086                                  GCA_156290       3         5609   \n",
       "912087                                  GCA_156290       3         5609   \n",
       "912088                                  GCA_156290       5         9951   \n",
       "912089                                  GCA_156290       5         9951   \n",
       "\n",
       "                                              peptidoform  exp_mass_to_charge  \\\n",
       "0                           AAAAAAAAAPAAAATAATTAATTAATAAQ         1172.107178   \n",
       "1                           AAAAAAAAAPAAAATAATTAATTAATAAQ         1172.107178   \n",
       "2                           AAAAAAAAAPAAAATAATTAATTAATAAQ         1172.107178   \n",
       "3         AAAFYKNILGAQ[Deamidated]VSEVVPLPEHGVSVVFVNLGNTK          918.255432   \n",
       "4                                             AAASHLFPFEK          406.549468   \n",
       "...                                                   ...                 ...   \n",
       "912085  YVQSAQSQ[Deamidated]IHNTC[Carbamidomethyl]WAMM...          900.753479   \n",
       "912086  YVSLENQHC[Carbamidomethyl]EHC[Carbamidomethyl]...          751.640869   \n",
       "912087  YVSLENQHC[Carbamidomethyl]EHC[Carbamidomethyl]...          751.640869   \n",
       "912088  YVSLENQ[Deamidated]HC[Carbamidomethyl]EHC[Carb...          517.421082   \n",
       "912089  YVSLENQ[Deamidated]HC[Carbamidomethyl]EHC[Carb...          517.421082   \n",
       "\n",
       "        calc_mass_to_charge                                  seq         tr  \\\n",
       "0               1171.600967        AAAAAAAAAPAAAATAATTAATTAATAAQ  4215.9185   \n",
       "1               1171.600967        AAAAAAAAAPAAAATAATTAATTAATAAQ  4215.9185   \n",
       "2               1171.600967        AAAAAAAAAPAAAATAATTAATTAATAAQ  4215.9185   \n",
       "3                917.997627  AAAFYKNILGAQVSEVVPLPEHGVSVVFVNLGNTK  5808.4260   \n",
       "4                406.548631                          AAASHLFPFEK  2901.2000   \n",
       "...                     ...                                  ...        ...   \n",
       "912085           900.411420              YVQSAQSQIHNTCWAMMGLMAVR  3952.7363   \n",
       "912086           751.642109                   YVSLENQHCEHCNSCTSK   823.0072   \n",
       "912087           751.642109                   YVSLENQHCEHCNSCTSK   823.0072   \n",
       "912088           517.214883                YVSLENQHCEHCNSCTSKDGR  1444.7499   \n",
       "912089           517.214883                YVSLENQHCEHCNSCTSKDGR  1444.7499   \n",
       "\n",
       "                                            modifications  ...  \\\n",
       "0                                                     NaN  ...   \n",
       "1                                                     NaN  ...   \n",
       "2                                                     NaN  ...   \n",
       "3                                           12|Deamidated  ...   \n",
       "4                                                     NaN  ...   \n",
       "...                                                   ...  ...   \n",
       "912085       8|Deamidated|13|Carbamidomethyl|17|Oxidation  ...   \n",
       "912086  9|Carbamidomethyl|12|Carbamidomethyl|15|Carbam...  ...   \n",
       "912087  9|Carbamidomethyl|12|Carbamidomethyl|15|Carbam...  ...   \n",
       "912088  7|Deamidated|9|Carbamidomethyl|12|Carbamidomet...  ...   \n",
       "912089  7|Deamidated|9|Carbamidomethyl|12|Carbamidomet...  ...   \n",
       "\n",
       "                 sample            protein               gene  \\\n",
       "0       GCA_009914755.4  ENSP05220080001.1  ENSG05220046501.1   \n",
       "1       GCA_009914755.4  ENSP05220080004.1  ENSG05220046501.1   \n",
       "2       GCA_009914755.4  ENSP05220080007.1  ENSG05220046501.1   \n",
       "3       GCA_009914755.4  ENSP05220077422.1  ENSG05220045157.1   \n",
       "4       GCA_009914755.4  ENSP05220065711.1  ENSG05220038055.1   \n",
       "...                 ...                ...                ...   \n",
       "912085  GCA_021951015.1  ENSP05525043594.1  ENSG05525025770.1   \n",
       "912086  GCA_021951015.1  ENSP05525070939.1  ENSG05525041216.1   \n",
       "912087  GCA_021951015.1  ENSP05525070944.1  ENSG05525041216.1   \n",
       "912088  GCA_021951015.1  ENSP05525070939.1  ENSG05525041216.1   \n",
       "912089  GCA_021951015.1  ENSP05525070944.1  ENSG05525041216.1   \n",
       "\n",
       "               transcript    gene_biotype  gene_symbol  \\\n",
       "0       ENST05220179403.1  protein_coding        SRP14   \n",
       "1       ENST05220179408.1  protein_coding        SRP14   \n",
       "2       ENST05220179411.1  protein_coding        SRP14   \n",
       "3       ENST05220173529.1  protein_coding         MCEE   \n",
       "4       ENST05220146556.1  protein_coding          XPC   \n",
       "...                   ...             ...          ...   \n",
       "912085  ENST05525097559.1  protein_coding          LSS   \n",
       "912086  ENST05525158826.1  protein_coding       ZCCHC4   \n",
       "912087  ENST05525158838.1  protein_coding       ZCCHC4   \n",
       "912088  ENST05525158826.1  protein_coding       ZCCHC4   \n",
       "912089  ENST05525158838.1  protein_coding       ZCCHC4   \n",
       "\n",
       "                                    gtf_index  \\\n",
       "0        ENSG05220046501,SRP14,protein_coding   \n",
       "1        ENSG05220046501,SRP14,protein_coding   \n",
       "2        ENSG05220046501,SRP14,protein_coding   \n",
       "3         ENSG05220045157,MCEE,protein_coding   \n",
       "4          ENSG05220038055,XPC,protein_coding   \n",
       "...                                       ...   \n",
       "912085     ENSG05525025770,LSS,protein_coding   \n",
       "912086  ENSG05525041216,ZCCHC4,protein_coding   \n",
       "912087  ENSG05525041216,ZCCHC4,protein_coding   \n",
       "912088  ENSG05525041216,ZCCHC4,protein_coding   \n",
       "912089  ENSG05525041216,ZCCHC4,protein_coding   \n",
       "\n",
       "                                             xref_index  \\\n",
       "0       ENSG05220046501,ENST05220179403,ENSP05220080001   \n",
       "1       ENSG05220046501,ENST05220179408,ENSP05220080004   \n",
       "2       ENSG05220046501,ENST05220179411,ENSP05220080007   \n",
       "3       ENSG05220045157,ENST05220173529,ENSP05220077422   \n",
       "4       ENSG05220038055,ENST05220146556,ENSP05220065711   \n",
       "...                                                 ...   \n",
       "912085  ENSG05525025770,ENST05525097559,ENSP05525043594   \n",
       "912086  ENSG05525041216,ENST05525158826,ENSP05525070939   \n",
       "912087  ENSG05525041216,ENST05525158838,ENSP05525070944   \n",
       "912088  ENSG05525041216,ENST05525158826,ENSP05525070939   \n",
       "912089  ENSG05525041216,ENST05525158838,ENSP05525070944   \n",
       "\n",
       "                                                 gtf_info  \\\n",
       "0       15,ensembl,gene,37841027,37844451,.,-,.,1,ense...   \n",
       "1       15,ensembl,gene,37841027,37844451,.,-,.,1,ense...   \n",
       "2       15,ensembl,gene,37841027,37844451,.,-,.,1,ense...   \n",
       "3       2,ensembl,gene,71120697,71141279,.,-,.,1,ensem...   \n",
       "4       3,ensembl,gene,14147095,14180485,.,-,.,1,ensem...   \n",
       "...                                                   ...   \n",
       "912085  21,ensembl,gene,38336946,38377856,.,-,.,1,ense...   \n",
       "912086  4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...   \n",
       "912087  4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...   \n",
       "912088  4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...   \n",
       "912089  4,ensembl,gene,25451616,25509248,.,+,.,1,ensem...   \n",
       "\n",
       "                                                xref_info  \n",
       "0       GO:0005786,GO:0005786,signal recognition parti...  \n",
       "1       GO:0005786,GO:0005786,signal recognition parti...  \n",
       "2       GO:0005786,GO:0005786,signal recognition parti...  \n",
       "3        UPI0000369399,UPI0000369399,,UniParc,CHECKSUM,,,  \n",
       "4       A0A024R2M8,A0A024R2M8,,UniProtKB generic acces...  \n",
       "...                                                   ...  \n",
       "912085  GO:0005811,GO:0005811,lipid droplet,GO,DEPENDE...  \n",
       "912086  GO:0003676,GO:0003676,nucleic acid binding,GO,...  \n",
       "912087  GO:0003676,GO:0003676,nucleic acid binding,GO,...  \n",
       "912088  GO:0003676,GO:0003676,nucleic acid binding,GO,...  \n",
       "912089  GO:0003676,GO:0003676,nucleic acid binding,GO,...  \n",
       "\n",
       "[912090 rows x 47 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_table(\"/mnt/nfs/wangd/project/paper_db/add_gtf_xref_info.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a139313",
   "metadata": {},
   "source": [
    "### gtf test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "263700df",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv('/mnt/nfs/wangd/project/paper_db/gtf/Homo_sapiens-GCA_018505865.1-2022_07-genes.gtf', sep='\\t', comment='#', header=None, names=['chrom', 'source','feature','start','end','score ','strand','frame','attribute'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "313b18dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf['gene_id'] = gtf['attribute'].str.extract('gene_id \"(.*?)\";')\n",
    "gtf['gene_version'] = gtf['attribute'].str.extract('gene_version \"(.*?)\";')\n",
    "gtf['gene_name'] = gtf['attribute'].str.extract('gene_name \"(.*?)\";')\n",
    "gtf['gene_source'] = gtf['attribute'].str.extract('gene_source \"(.*?)\";')\n",
    "gtf['gene_biotype'] = gtf['attribute'].str.extract('gene_biotype \"(.*?)\";')\n",
    "gtf['projection_parent_gene'] = gtf['attribute'].str.extract('projection_parent_gene \"(.*?)\";')\n",
    "\n",
    "gtf.fillna(\"\",inplace=True)\n",
    "gtf.index = gtf[\"gene_id\"] + \",\" + gtf[\"gene_name\"] + \",\" + gtf[\"gene_biotype\"]\n",
    "\n",
    "gtf.drop(['attribute', 'gene_id', 'gene_name', 'gene_biotype'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "b5a2c05f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3100070it [01:39, 31058.78it/s]\n"
     ]
    }
   ],
   "source": [
    "gtf_dict = {}\n",
    "for index, row in tqdm(gtf.iterrows()):\n",
    "    if index in gtf_dict:\n",
    "        gtf_dict[index].append( ','.join(map(str, row)))\n",
    "    else:\n",
    "        gtf_dict[index] = [ ','.join(map(str, row))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa031e53",
   "metadata": {},
   "source": [
    "### xref test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "92795989",
   "metadata": {},
   "outputs": [],
   "source": [
    "xref = pd.read_table(\"/mnt/nfs/wangd/project/paper_db/xref/Homo_sapiens-GCA_018852615.1-2022_07-xref.tsv\")\n",
    "\n",
    "xref.fillna(\"\",inplace=True)\n",
    "xref.index = xref[\"gene_stable_id\"] + \",\" + xref[\"transcript_stable_id\"] + \",\" + xref[\"protein_stable_id\"]\n",
    "xref.drop(['gene_stable_id', 'transcript_stable_id', 'protein_stable_id'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "ae5ed951",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "848450it [00:27, 30632.87it/s]\n"
     ]
    }
   ],
   "source": [
    "xref_dict = {}\n",
    "for index, row in tqdm(xref.iterrows()):\n",
    "    if index in xref_dict:\n",
    "        xref_dict[index].append( ','.join(map(str, row)))\n",
    "    else:\n",
    "        xref_dict[index] = [ ','.join(map(str, row))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4640be0f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
